crml/nft/src/lib.rs

```diff
@@ -0,0 +1,651 @@
+		/// Winning bids on open listings. keyed by collection id and token id
+		pub ListingWinningBid get(fn listing_winning_bid): double_map hasher(blake2_128_concat) CollectionId, hasher(twox_64_concat) T::TokenId => Option<(T::AccountId, Balance)>;
+		/// Block numbers where listings will close. It is `Some` if at block number, (collection id, token id) is listed and scheduled to close.
+		pub ListingEndSchedule get(fn listing_end_schedule): double_map hasher(twox_64_concat) T::BlockNumber, hasher(twox_64_concat) (CollectionId, T::TokenId) => Option<()>;
+	}
+}
+
+/// The maximum number of attributes in an NFT collection schema
+pub const MAX_SCHEMA_FIELDS: u32 = 16;
+/// The maximum length of valid collection IDs
+pub const MAX_COLLECTION_ID_LENGTH: u8 = 32;
+
+pub(crate) const LOG_TARGET: &str = "nft";
+
+// syntactic sugar for logging.
+#[macro_export]
+macro_rules! log {
+	($level:tt, $patter:expr $(, $values:expr)* $(,)?) => {
+		frame_support::debug::$level!(
+			target: crate::LOG_TARGET,
+			$patter $(, $values)*
+		)
+	};
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin, system = frame_system {
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		/// Check and close all expired listings
+		fn on_initialize(now: T::BlockNumber) -> Weight {
+			let removed_count = Self::close_listings_at(now);
+			// 'direct_purchase' weight is comparable to succesful closure of an auction
+			T::WeightInfo::direct_purchase() * removed_count as Weight
+		}
+
+		/// Create a new NFT collection
+		/// The caller will be come the collection' owner
+		/// `collection_id`- 32 byte utf-8 string
+		/// `schema` - for the collection
+		/// `royalties_schedule` - defacto royalties plan for secondary sales, this will apply to all tokens in the collection by default.
+		#[weight = T::WeightInfo::create_collection()]
+		fn create_collection(origin, collection_id: CollectionId, schema: NFTSchema, royalties_schedule: Option<RoyaltiesSchedule<T::AccountId>>) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+
+			ensure!(!collection_id.is_empty() && collection_id.len() <= MAX_COLLECTION_ID_LENGTH as usize, Error::<T>::CollectionIdInvalid);
+			ensure!(core::str::from_utf8(&collection_id).is_ok(), Error::<T>::CollectionIdInvalid);
+			ensure!(!<CollectionOwner<T>>::contains_key(&collection_id), Error::<T>::CollectionIdExists);
+
+			ensure!(!schema.is_empty(), Error::<T>::SchemaEmpty);
+			ensure!(schema.len() <= MAX_SCHEMA_FIELDS as usize, Error::<T>::SchemaMaxAttributes);
+
+			let mut set = BTreeSet::new();
+			for (name, type_id) in schema.iter() {
+				// Check the provided attribute types are valid
+				ensure!(NFTAttributeValue::is_valid_type_id(*type_id), Error::<T>::SchemaInvalid);
+				// Attribute names must be unique (future proofing for map lookups etc.)
+				ensure!(set.insert(name), Error::<T>::SchemaDuplicateAttribute);
```

Comment:
I mentioned this on the previous revision of this file: I think we should allow the names for the attributes to be optional as a token schema creator may want to rely solely on the order of the fields and give a name to them in the application. They can save another bit of storage by that. If we do so, then we should handle the default or empty attribute name here as well.

Reply:
Makes it too complicated for edge case at this stage, if it turns out this would be useful we can readdress it